src/VecSim/algorithms/hnsw/hnsw.h

```diff
@@ -1220,51 +1303,44 @@ int HNSWIndex<DataType, DistType>::appendVector(const void *vector_data, const l
         if (element_max_level < maxlevelcopy) {
-            DistType cur_dist =
-                this->dist_func(vector_data, getDataByInternalId(currObj), this->dim);
+            cur_dist = this->dist_func(vector_data, getDataByInternalId(currObj), this->dim);
             for (size_t level = maxlevelcopy; level > element_max_level; level--) {
                 // this is done for the levels which are above the max level
                 // to which we are going to insert the new element. We do
                 // a greedy search in the graph starting from the entry point
                 // at each level, and move on with the closest element we can find.
                 // When there is no improvement to do, we take a step down.
-                bool changed = true;
-                while (changed) {
-                    changed = false;
-                    unsigned int *data;
-#ifdef ENABLE_PARALLELIZATION
-                    std::unique_lock<std::mutex> lock(link_list_locks_[currObj]);
-#endif
-                    data = get_linklist(currObj, level);
-                    int size = getListCount(data);
-
-                    auto *datal = (idType *)(data + 1);
-                    for (int i = 0; i < size; i++) {
-                        idType cand = datal[i];
-                        if (cand < 0 || cand > max_elements_)
-                            throw std::runtime_error(
-                                "candidate error: candidate id is out of index range");
-
-                        DistType d =
-                            this->dist_func(vector_data, getDataByInternalId(cand), this->dim);
-                        if (d < cur_dist) {
-                            cur_dist = d;
-                            currObj = cand;
-                            changed = true;
-                        }
-                    }
-                }
+                greedilySearchLevel<false>(vector_data, level, currObj, cur_dist);
             }
         }
 
-        for (size_t level = std::min(element_max_level, maxlevelcopy); (int)level >= 0; level--) {
-            if (level > maxlevelcopy || level < 0) // possible?
-                throw std::runtime_error("Level error");
+        auto max_common_level = std::min(element_max_level, maxlevelcopy);
+        if (this->num_marked_deleted) {
+            if (cur_dist == std::numeric_limits<DistType>::max()) {
+                cur_dist = this->dist_func(vector_data, getDataByInternalId(currObj), this->dim);
+            }
+            for (size_t level = max_common_level; (int)level >= 0; level--) {
+
+                candidatesMaxHeap<DistType> top_candidates =
+                    searchLayer<true>(currObj, vector_data, level, ef_construction_);
+                if (top_candidates.empty()) {
+                    greedilySearchLevel<false>(vector_data, level, currObj, cur_dist);
+                    setIncomingEdgesPtr(cur_c, level,
```

Comment:
does this means this is the only valid node in the level? do we have a test for this?

Reply:
1. yes
2. no, will add